doc/ip/next_hop_group_hld.md

```diff
@@ -0,0 +1,148 @@
+
+If a next hop group cannot be programmed because the data plane limit has been reached, one next hop will be picked to be temporarily used for that group.  When a route is then programmed using that next hop group, it will be informed that a temporary form of the next hop group is in use, and so the route will remain on the pending list and will retry until the correct next hop group is programmed.  This mirrors the current behaviour in the route orchagent when the next hop group limit is reached.
+
+A few notes about this design:
+
+- The next hop groups created by the next hop group orchagent will not interact with those created by the route orchagent's current mechanisms. This is because the route orchagent's groups are identified by the members of the group, and so changing those members makes it a different group, whereas the new next hop group orchagent's groups have an arbitrary index, with members able to change without it becoming a different group. Therefore even if a group in each category share the same members, they will be programmed separately into ASIC_DB, in case the members change. The expectation is that all routes will be programmed into APP_DB using the old next hop schema or the new one, and so the only occasion where next hop groups could overlap is if a next hop group is created for an ACL that overlaps with one created by the next hop group orchagent.
+- A ROUTE_TABLE entry with both next hop information embedded and a reference to the NEXT_HOP_GROUP_TABLE will be ignored.
+- Next hop groups created by the fine grained next hop group orchagent will be unchanged by this design.
+
+## 3.4 SyncD
+No change.
+
+
+## 3.5 SAI
+No changes are being made in SAI. The end result of what gets programmed over the SAI will be the same as currently gets programmed - the difference is the programming sequence to get there.
+
+## 3.6 CLI
+
+### 3.6.1 Show Commands
+
+The output of 'show ip route' and 'show ipv6 route' will remain unchanged - the CLI code will resolve the next hop group ID referenced in the ROUTE_TABLE to display the next hops for the routes.
+
+# 4 Warm Boot Support
+
+Warm boot support must not be affected by this enhancement. Specifically, this means that (assuming the required routes don't change during warm boot), routes must be associated with the same next hop group after warm boot. The application programming APP_DB must therefore ensure that next hop groups get assigned the same identifier after reboot. This could be done by the application maintaining that information itself, or retrieving the identifiers from APP_DB at the start of warm boot.
+
+# 4.1 Warm Upgrade
+
+Existing applications will not be affected by this enhancement, so no warm upgrade support is needed.
```

Comment:
>so no warm upgrade support is needed [](start = 64, length = 36)

If warm upgrade from old image to new image, there is no next hop group table entries at all before warm-reboot doTask(). How do you handle ?

Reply:
I am asking warm upgrade(reboot) from old image (without your feature) to new image (with your feature). If you don't handle the DB migration, the warm reboot will fail, and there will be data plane disrutpion.

Reply:
Since no existing application is being changed to use the new feature, I don't think it makes sense to add warm upgrade support now, as it wouldn't be testable. Warm upgrade support should be added at the point it becomes needed. I've updated the doc to this effect.